src/charm.py

```diff
@@ -88,13 +90,33 @@ def __init__(self, *args):
             ],
             network_attachment_definitions_func=self._network_attachment_definitions_from_config,
         )
+        self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.bessd_pebble_ready, self._on_bessd_pebble_ready)
         self.framework.observe(self.on.pfcp_agent_pebble_ready, self._on_pfcp_agent_pebble_ready)
         self.framework.observe(
             self.fiveg_n3_provider.on.fiveg_n3_request, self._on_fiveg_n3_request
         )
 
+    def _on_start(self, event: EventBase) -> None:
+        """Handler for Juju start event.
+
+        This handler enforces usage of a CPU which supports instructions required to run this
+        charm. If the CPU doesn't meet the requirements, charm goes to Blocked state.
+
+        Args:
+            event: Juju event
+        """
+        if not self._is_cpu_compatible():
+            self.unit.status = BlockedStatus("CPU is not compatible")
```

Comment:
Isn't error status more fitting?
The user won't be able to take any action to recover from Blocked anyway.
And without this check the charm will end up in error.
(But again, is it an error of the charm? probably not :D )

Reply:
Yes, I agree, but the problem with the `ErrorStatus` is that it can't be set by the charm. 
We could raise an exception here to put charm in `ErrorStatus`, but in that case I'm not sure we can set the message in `juju status`.

Reply:
Ah okay I see.
In that case it's all good. :)

Reply:
It would appear we need to be more aggressive here.  The message only shows in the debug log, and it's hard to find:
```
unit-sdcore-upf-0: 13:07:52 INFO juju.worker.uniter found queued "install" hook
unit-sdcore-upf-0: 13:07:54 INFO unit.sdcore-upf/0.juju-log Running legacy hooks/install.
unit-sdcore-upf-0: 13:07:55 WARNING unit.sdcore-upf/0.juju-log 2 containers are present in metadata.yaml and refresh_event was not specified. Defaulting to update_status. Metrics IP may not be set in a timely fashion.
unit-sdcore-upf-0: 13:07:55 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/api/v1/namespaces/core/services/sdcore-upf "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:07:55 INFO unit.sdcore-upf/0.juju-log HTTP Request: PATCH https://10.152.183.1/api/v1/namespaces/core/services/sdcore-upf "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:07:55 INFO unit.sdcore-upf/0.juju-log Kubernetes service 'sdcore-upf' patched successfully
unit-sdcore-upf-0: 13:07:56 ERROR unit.sdcore-upf/0.juju-log CPU is not compatible!
Please use a CPU that has the following capabilities: avx2, rdrand
unit-sdcore-upf-0: 13:07:56 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-sdcore-upf-0: 13:07:56 INFO juju.worker.uniter found queued "leader-elected" hook
unit-sdcore-upf-0: 13:07:57 WARNING unit.sdcore-upf/0.juju-log 2 containers are present in metadata.yaml and refresh_event was not specified. Defaulting to update_status. Metrics IP may not be set in a timely fashion.
unit-sdcore-upf-0: 13:07:57 ERROR unit.sdcore-upf/0.juju-log CPU is not compatible!
Please use a CPU that has the following capabilities: avx2, rdrand
unit-sdcore-upf-0: 13:07:57 INFO juju.worker.uniter.operation ran "leader-elected" hook (via hook dispatching script: dispatch)
unit-sdcore-upf-0: 13:07:58 WARNING unit.sdcore-upf/0.juju-log 2 containers are present in metadata.yaml and refresh_event was not specified. Defaulting to update_status. Metrics IP may not be set in a timely fashion.
unit-sdcore-upf-0: 13:07:59 ERROR unit.sdcore-upf/0.juju-log CPU is not compatible!
Please use a CPU that has the following capabilities: avx2, rdrand
unit-sdcore-upf-0: 13:07:59 INFO juju.worker.uniter.operation ran "pfcp-agent-pebble-ready" hook (via hook dispatching script: dispatch)
unit-sdcore-upf-0: 13:08:00 WARNING unit.sdcore-upf/0.juju-log 2 containers are present in metadata.yaml and refresh_event was not specified. Defaulting to update_status. Metrics IP may not be set in a timely fashion.
unit-sdcore-upf-0: 13:08:00 ERROR unit.sdcore-upf/0.juju-log CPU is not compatible!
Please use a CPU that has the following capabilities: avx2, rdrand
unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/k8s.cni.cncf.io/v1/namespaces/core/network-attachment-definitions/access-net "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/k8s.cni.cncf.io/v1/namespaces/core/network-attachment-definitions/core-net "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/apps/v1/namespaces/core/statefulsets/sdcore-upf "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/api/v1/namespaces/core/pods/sdcore-upf-0 "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log Default core network route created
unit-sdcore-upf-0: 13:08:00 ERROR unit.sdcore-upf/0.juju-log iptables: Bad rule (does a matching rule exist in that chain?).

unit-sdcore-upf-0: 13:08:00 INFO unit.sdcore-upf/0.juju-log Iptables rule for ICMP created
unit-sdcore-upf-0: 13:08:02 INFO unit.sdcore-upf/0.juju-log Service `routectl` restarted
unit-sdcore-upf-0: 13:08:02 ERROR unit.sdcore-upf/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/./src/charm.py", line 690, in <module>
    main(UPFOperatorCharm)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/framework.py", line 344, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/framework.py", line 833, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/framework.py", line 922, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/./src/charm.py", line 264, in _on_bessd_pebble_ready
    self._configure_bessd_workload()
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/./src/charm.py", line 309, in _configure_bessd_workload
    self._bessd_container.restart(self._bessd_service_name)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/model.py", line 1909, in restart
    self._pebble.restart_services(service_names)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/pebble.py", line 1718, in restart_services
    return self._services_action('restart', services, timeout, delay)
  File "/var/lib/juju/agents/unit-sdcore-upf-0/charm/venv/ops/pebble.py", line 1739, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "bessd" (cannot start service: exited quickly with code 132)
----- Logs from task 0 -----
2023-09-08T13:08:02Z INFO Service "bessd" has never been started.
----- Logs from task 1 -----
2023-09-08T13:08:02Z INFO Most recent service output:
    
2023-09-08T13:08:02Z ERROR cannot start service: exited quickly with code 132
-----
unit-sdcore-upf-0: 13:08:02 ERROR juju.worker.uniter.operation hook "bessd-pebble-ready" (via hook dispatching script: dispatch) failed: exit status 1
unit-sdcore-upf-0: 13:08:02 ERROR juju.worker.uniter pebble poll failed for container "bessd": failed to send pebble-ready event: hook failed
unit-sdcore-upf-0: 13:08:03 WARNING unit.sdcore-upf/0.juju-log 2 containers are present in metadata.yaml and refresh_event was not specified. Defaulting to update_status. Metrics IP may not be set in a timely fashion.
unit-sdcore-upf-0: 13:08:03 ERROR unit.sdcore-upf/0.juju-log CPU is not compatible!
Please use a CPU that has the following capabilities: avx2, rdrand
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/k8s.cni.cncf.io/v1/namespaces/core/network-attachment-definitions/access-net "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/k8s.cni.cncf.io/v1/namespaces/core/network-attachment-definitions/core-net "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/apis/apps/v1/namespaces/core/statefulsets/sdcore-upf "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log HTTP Request: GET https://10.152.183.1/api/v1/namespaces/core/pods/sdcore-upf-0 "HTTP/1.1 200 OK"
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log Default core network route created
unit-sdcore-upf-0: 13:08:04 INFO unit.sdcore-upf/0.juju-log Starting configuration of the `bessd` service
unit-sdcore-upf-0: 13:08:05 ERROR unit.sdcore-upf/0.juju-log *** Error: BESS daemon not connected

unit-sdcore-upf-0: 13:08:05 ERROR unit.sdcore-upf/0.juju-log   Command failed: run /opt/bess/bessctl/conf/up4

unit-sdcore-upf-0: 13:08:05 INFO unit.sdcore-upf/0.juju-log Failed running configuration for bess
unit-sdcore-upf-0: 13:08:07 INFO unit.sdcore-upf/0.juju-log Starting configuration of the `bessd` service
unit-sdcore-upf-0: 13:08:07 ERROR unit.sdcore-upf/0.juju-log *** Error: BESS daemon not connected

unit-sdcore-upf-0: 13:08:07 ERROR unit.sdcore-upf/0.juju-log   Command failed: run /opt/bess/bessctl/conf/up4

unit-sdcore-upf-0: 13:08:07 INFO unit.sdcore-upf/0.juju-log Failed running configuration for bess
```

Juju status shows:
```
Model  Controller          Cloud/Region        Version  SLA          Timestamp
core   microk8s-localhost  microk8s/localhost  3.1.5    unsupported  13:10:52Z

App         Version  Status   Scale  Charm          Channel  Rev  Address        Exposed  Message
router               active       1  sdcore-router             1  10.152.183.81  no       
sdcore-upf           waiting      1  sdcore-upf                1  10.152.183.80  no       installing agent

Unit           Workload  Agent      Address      Ports  Message
router/0*      active    idle       10.1.150.17         
sdcore-upf/0*  waiting   executing  10.1.150.22         Waiting for bessd service to run
```

Reply:
My bad. `BlockedStatus` doesn't stop other events from being fired. Looks like raising an Exception is the way to go after all.